### ISSUE TITLE
[Suggestion] failed library search regex doesn't search again if a new character is added

### ISSUE THREAD

**Issue by djvs** [arybczak/ncmpcpp#66]
If no matches found for string:

    Darbar

don't search again for:

    Darbare

Maybe a good idea... sounds simple to do

**Comment by arybczak** [comment]
Is that a problem?

**Comment by djvs** [comment]
On a pretty modern computer it's maybe 0.015ms per song, which adds up fast (2 seconds per character on a 150k library)

**Comment by arybczak** [comment]
Hmm, you need to specify the context in which you experience this problem.

**Comment by djvs** [comment]
Normal use, although I believe it's operating over a socket on localhost?

**Comment by djvs** [comment]
Normally it's quite responsive though, so...

**Comment by arybczak** [comment]
Yeah, but what is the exact situation (in other words, steps to reproduce)?

**Comment by djvs** [comment]
When entering a search regex on a  library with a large amount of songs (>150k), entering the wrong character and then continuing to type can result in a lengthy delay as library continues to be searched for an increasingly specific regex that ncmpcpp could easily predict not to match, given that the last search string (contained as a subset in the current search string) did not match.  Therefore it could skip the search until the next search which no longer contained the first non-matching search string.   Like if I'm looking for `Chopin Preludes`, and I type `Chopon`, each of the next 9 characters entered in ` Preludes` will result in an unsuccessful search of the entire library.

**Comment by arybczak** [comment]
You are assuming that the following is true: `Let D be a set of strings. If S is a regular expression that doesn't match any of the strings from D, then for all strings T a regular expression ST also doesn't match any of the strings from D.` Unfortunately it's not, consider D = {"a"}, S = "b", T = "|a".

**Comment by djvs** [comment]
ncmpcpp supports extended regex?

Anyway, you could code in those exceptions, no?

**Comment by arybczak** [comment]
It supports extended and perl.

An no, I couldn't, because that's just one I cant think of and there are possibly more.

**Comment by djvs** [comment]
Hm.. is there a conf option for disabling the auto-find then?

**Comment by arybczak** [comment]
No, although I guess I could add it.

**Comment by djvs** [comment]
Perhaps you could disable the match if any of the special characters are present in the pre-regex search string: `.[{}()\*+?|^$`.

**Comment by djvs** [comment]
It's a little tricky but you gotta admit it'd be a handy feature for large libraries.

**Comment by czipperz** [comment]
@djvs I was a bit interested so I looked into the source.

Seems like you would need to rewrite the readline library, which would be a huge pain in the ass.
See window.cpp -> Window::prompt

I assume you using `Browse => Find forward` for searching.

**Comment by arybczak** [comment]
@czipperz  You can plug hooks into it, that's what makes the immediate search work.

@djvs I'm thinking that it could be split into two modes, the literal and regex search. Then what you propose could be done in literal mode.
